src/librustc_target/abi/call/wasm32.rs

```diff
@@ -1,16 +1,38 @@
-use crate::abi::call::{FnType, ArgType};
+use crate::abi::call::{FnType, ArgType, Uniform};
+use crate::abi::{HasDataLayout, LayoutOf, TyLayout, TyLayoutMethods};
 
-fn classify_ret_ty<Ty>(ret: &mut ArgType<'_, Ty>) {
-    ret.extend_integer_width_to(32);
+fn classify_ret_ty<'a, Ty, C>(cx: &C, ret: &mut ArgType<'a, Ty>)
+    where Ty: TyLayoutMethods<'a, C> + Copy,
+          C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>> + HasDataLayout
+{
+    if ret.layout.is_aggregate() {
+        if let Some(unit) = ret.layout.homogeneous_aggregate(cx).unit() {
+            let size = ret.layout.size;
+            if unit.size == size {
+                ret.cast_to(Uniform {
+                    unit,
+                    total: size
+                });
+                return;
+            }
+        }
+
+        ret.make_indirect();
+    }
 }
 
 fn classify_arg_ty<Ty>(arg: &mut ArgType<'_, Ty>) {
-    arg.extend_integer_width_to(32);
```

Comment:
Same here.

Reply:
I agree with @bjorn3, Clang appears to do this: https://github.com/llvm/llvm-project/blob/d01b4a786271c9ce9df43add56ec2e3fb40ce920/clang/lib/CodeGen/TargetInfo.cpp#L701

cc @rkruppe @nagisa @nikic

Reply:
Yeah, it's confusing that this happens in both the frontend and the backend. I suppose if we're trying to match clang here then it would make sense to do the expansion, although practically it shouldn't make a difference.

Reply:
I think the difference is we pick sign- or zero- extension *depending on the type*, which LLVM doesn't know, so presumably it would be correct for unsigned integers but not signed, or vice-versa.

I didn't really like this entire dance but IIRC @rkruppe explained at some point that it makes life easier for architectures with e.g. 32-bit registers that you want to do 16-bit or 8-bit operations within, if everything is appropriately extended with its individual signedness.

Reply:
Yeah that makes sense actually :+1: